### ISSUE TITLE
iscsiadm failed with ipv6 when discovery and login

### ISSUE THREAD

**Issue by YanfeiXu** [open-iscsi/open-iscsi#400]
Hi,
I am encountering some failures of iscsiadm when using ipv6, but it is good with ipv4.  Details as below:

dracut:/# iscsiadm --version
iscsiadm version 6.2.1.4-1

dracut:/# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
    link/ether 00:02:00:00:03:18 brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.2/24 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::202:ff:fe00:318/64 scope link noprefixroute
       valid_lft forever preferred_lft forever

dracut:/# ping fe80::202:ff:fe00:318
PING fe80::202:ff:fe00:318(fe80::202:ff:fe00:318) 56 data bytes
64 bytes from fe80::202:ff:fe00:318%eth0: icmp_seq=1 ttl=64 time=0.039 ms
64 bytes from fe80::202:ff:fe00:318%eth0: icmp_seq=2 ttl=64 time=0.009 ms

dracut:/# iscsiadm  -m discovery -t sendtargets -p [fe80:0000:0000:0000:0203:00ff:fe00:319]:3260
iscsiadm: cannot make connection to fe80::203:ff:fe00:319: Invalid argument
iscsiadm: cannot make connection to fe80::203:ff:fe00:319: Invalid argument
iscsiadm: cannot make connection to fe80::203:ff:fe00:319: Invalid argument
iscsiadm: cannot make connection to fe80::203:ff:fe00:319: Invalid argument

dracut:/# iscsiadm -m node -T iqn.e2000:acc -l -p [fe80::203:ff:fe00:319]:3260
iscsiadm: No records found

After convert it to ipv4 ip address, everything goes fine. I found this issue when using ibft with ipv6 setting at the beginning.

Could you give some suggestions? Thanks!

**Comment by gonzoleeman** [comment]
Hi:

It looks like you are trying to use an IPv6 link-local address. These are not routable, FYI.

> 2: ens33: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
>     link/ether 00:0c:29:52:fb:54 brd ff:ff:ff:ff:ff:ff
>     altname enp2s1
>     inet 192.168.10.32/24 brd 192.168.10.255 scope global dynamic noprefixroute ens33
>        valid_lft 85569sec preferred_lft 85569sec
>     inet6 2601:1c0:4d7f:138e::12/128 scope global dynamic noprefixroute 
>        valid_lft 2774sec preferred_lft 2774sec
>     inet6 fe80::c1bc:d8c1:e3da:7610/64 scope link noprefixroute 
>        valid_lft forever preferred_lft forever
> 

On my system, if I use:

> ***@***.***(1):~# ping -c1 -W1 'fe80::c1bc:d8c1:e3da:7610'      
> ping: Warning: IPv6 link-local address on ICMP datagram socket may require ifname or scope-id => use: address%<ifname|scope-id>
> PING fe80::c1bc:d8c1:e3da:7610(fe80::c1bc:d8c1:e3da:7610) 56 data bytes
> 
> --- fe80::c1bc:d8c1:e3da:7610 ping statistics ---
> 1 packets transmitted, 0 received, 100% packet loss, time 0ms
> 
> ***@***.***(1):~# ping -c1 -W1 'fe80::c1bc:d8c1:e3da:7610%ens33'
> PING fe80::c1bc:d8c1:e3da:7610%ens33(fe80::c1bc:d8c1:e3da:7610%ens33) 56 data bytes
> 64 bytes from fe80::c1bc:d8c1:e3da:7610%ens33: icmp_seq=1 ttl=64 time=0.021 ms
> 
> --- fe80::c1bc:d8c1:e3da:7610%ens33 ping statistics ---
> 1 packets transmitted, 1 received, 0% packet loss, time 0ms
> rtt min/avg/max/mdev = 0.021/0.021/0.021/0.000 ms
> 

So ping only works if I qualify the link-local IPv6 address with the interface name. Same thing for iSCSI discovery:

> ***@***.***(1):~# iscsiadm -m discovery -t st -p 'fe80::c1bc:d8c1:e3da:7610'   
> iscsiadm: cannot make connection to fe80::c1bc:d8c1:e3da:7610: Invalid argument
> iscsiadm: cannot make connection to fe80::c1bc:d8c1:e3da:7610: Invalid argument
> iscsiadm: cannot make connection to fe80::c1bc:d8c1:e3da:7610: Invalid argument
> iscsiadm: cannot make connection to fe80::c1bc:d8c1:e3da:7610: Invalid argument
> iscsiadm: cannot make connection to fe80::c1bc:d8c1:e3da:7610: Invalid argument
> iscsiadm: cannot make connection to fe80::c1bc:d8c1:e3da:7610: Invalid argument
> iscsiadm: connection login retries (reopen_max) 5 exceeded
> iscsiadm: Could not perform SendTargets discovery: iSCSI PDU timed out
> ***@***.***(1):~# iscsiadm -m discovery -t st -p 'fe80::c1bc:d8c1:e3da:7610%ens33' 
> [fe80::c1bc:d8c1:e3da:7610]:3260,1 iqn.2003-01.org.linux-iscsi.linux-tw.x8664:sn.e29c473ac9d0
> [fe80::c1bc:d8c1:e3da:7610]:3260,1 iqn.2003-01.org.linux-iscsi.linux-tw.x8664:sn.6880d33c4104
> ***@***.***(1):~# 

And discovery also works to my global IPv6 address, without having to add the ‘%IFNAME’ qualifier.

> On Mar 5, 2023, at 8:40 PM, Yanfei Xu ***@***.***> wrote:
> 
> 
> Hi,
> I am encountering some failures of iscsiadm when using ipv6, but it is good with ipv4. Details as below:
> 
> dracut:/# iscsiadm --version
> iscsiadm version 6.2.1.4-1
> 
> dracut:/# ip a
> 1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
> link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
> inet 127.0.0.1/8 scope host lo
> valid_lft forever preferred_lft forever
> inet6 ::1/128 scope host
> valid_lft forever preferred_lft forever
> 2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP group default qlen 1000
> link/ether 00:02:00:00:03:18 brd ff:ff:ff:ff:ff:ff
> inet 192.168.0.2/24 scope global eth0
> valid_lft forever preferred_lft forever
> inet6 fe80::202:ff:fe00:318/64 scope link noprefixroute
> valid_lft forever preferred_lft forever
> 

Interesting that you don’t have a global scope IPv6 address on eth0.
> dracut:/# ping fe80::202:ff:fe00:318
> PING fe80::202:ff:fe00:318(fe80::202:ff:fe00:318) 56 data bytes
> 64 bytes from fe80::202:ff:fe00:318%eth0: icmp_seq=1 ttl=64 time=0.039 ms
> 64 bytes from fe80::202:ff:fe00:318%eth0: icmp_seq=2 ttl=64 time=0.009 ms
> 
> dracut:/# iscsiadm -m discovery -t sendtargets -p [fe80:0000:0000:0000:0203:00ff:fe00:319]:3260
> iscsiadm: cannot make connection to fe80::203:ff:fe00:319: Invalid argument
> iscsiadm: cannot make connection to fe80::203:ff:fe00:319: Invalid argument
> iscsiadm: cannot make connection to fe80::203:ff:fe00:319: Invalid argument
> iscsiadm: cannot make connection to fe80::203:ff:fe00:319: Invalid argument
> 
> dracut:/# iscsiadm -m node -T iqn.e2000:acc -l -p [fe80::203:ff:fe00:319]:3260
> iscsiadm: No records found
> 

You do not need the “:3260”, since that’s the default.
> After convert it to ipv4 ip address, everything goes fine. I found this issue when using ibft with ipv6 setting at the beginning.
> 
> Could you give some suggestions? Thanks!
> 

So, bottom line, use a global scope IPv6 address or add the ‘%IFNAME’ qualifier to your link-local addresses.

— 
Lee Duncan

**Comment by gonzoleeman** [comment]
I believe this issue can be closed, but I'll wait to hear back from you @YanfeiXu in case you have more questions or issues.

**Comment by YanfeiXu** [comment]
> I believe this issue can be closed, but I'll wait to hear back from you @YanfeiXu in case you have more questions or issues.

Adding the interface name and dropping port number are both working for me! Thanks for your suggestion! Now I remain a confusion about using iscsi with dracut, as I edited the issue content so you may didn't notice it in the second version, I paste it here: 
I am seeing the iscsiroot.sh of dracut always show warning message as below when using ibft with kernel cmdline "ip=ibft root=/dev/sda netroot=iscsi rd.iscsi.ibft=1 rd.iscsi.firmware=1 initrdmem=0x440c000000,0x8000000 acpi=force". According to man page of dracut, "netroot=iscsi" is suggested to be used with ibft, why can't the iscsiadm identify the parameter "iscsi:3260" generated by dracut script.

[ 61.293179] dracut Warning: Target discovery to iscsi:3260 failed with status 0
iscsiadm: Cannot resolve host iscsi. getaddrinfo error: [Name or service not known]
iscsiadm: cannot resolve host name iscsi
iscsiadm: cannot resolve host name iscsi
iscsiadm: Could not perform SendTargets discovery: encountered connection failure
dracut Warning: Target discovery to **iscsi:3260** failed with status 0

As you suggested 3260 should be dropped, then there should be some bugs in dracut? Or maybe I should put this dracut related question to dracut community.

**Comment by YanfeiXu** [comment]
> I believe this issue can be closed, but I'll wait to hear back from you @YanfeiXu in case you have more questions or issues.

May you help to answer the above question about "dracut+iscsi"? If there is nothing for iscsi to handle, I'd like to close this issue. :)

**Comment by gonzoleeman** [comment]
This is perhaps more of a question for your distribution, since each distro does iscsi root boot a little differently, in my experience. I know how SUSE does it pretty well :), but not so much the other distros.

I _believe_ that you are settings **netroot** incorrectly. For our distro, it is:

`netroot=iscsi:IP_ADDR:PROTOCOL:PORT:LUN:TARGETNAME
`
Where **PROTOCOL** is usually left blank, taking the default.

Looking at the top of parse-iscsiroot.sh in modules.d/95iscsi, I see a comment that says:

```
# Preferred format:
#       root=iscsi:[<servername>]:[<protocol>]:[<port>]:[<LUN>]:<targetname>
#       [root=*] netroot=iscsi:[<servername>]:[<protocol>]:[<port>]:[<LUN>]:<targetname>
#
# Legacy formats:
#       [net]root=[iscsi] iscsiroot=[<servername>]:[<protocol>]:[<port>]:[<LUN>]:<targetname>
#       [net]root=[iscsi] iscsi_firmware
#
# root= takes precedence over netroot= if root=iscsi[...]
```

To me, this says:

- You do not need rd.iscsi.firmware (unless you are booting from a Converged Network Adapter (in which case you wouldn't be using iBFT)
- You are setting **netroot** incorrectly
- You are overriding **netroot** with your **root=**, so remove that

I hope this helps.

**Comment by gonzoleeman** [comment]
Note: I'm not positive at all about the need to remove "root=...", so try that, and if it does't work, put it back?

**Comment by gonzoleeman** [comment]
@YanfeiXu is your issue resolved now?

**Comment by YanfeiXu** [comment]
> @YanfeiXu is your issue resolved now?

Sorry for late reply. What I want is to use IBFT on our platform with IPv6. I was trying but failed. Then I want to verify this step by step. First, verify if the static IPv6 setting in cmdline works for iscsiboot (as you see, I created this issue). Second, verify on using IBFT with IPV6.  Agree with your comment about "root=", it is no need a/c  parse-iscsiroot.sh.  Below is the experiment result for static IPv6 setting:
[    0.000000] Kernel command line: ip=[fe80::202:ff:fe00:318]::[fe80::202:ff:fe00:319]:64::eth0:none root=iscsi:[fe80::202:ff:fe00:319%eth0]::::iqn.e2000:acc rd.break initrdmem=0x440C
000000,0x8000000
....
Target discovery to fe80::202:ff:fe00:319%eth0:3260 failed with status 0
[  133.050468] random: NetworkManager: uninitialized urandom read (16 bytes read)
[  133.146796] random: ln: uninitialized urandom read (6 bytes read)
iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
iscsiadm: connection login retries (reopen_max) 5 exceeded
iscsiadm: Could not perform SendTargets discovery: iSCSI PDU timed out
Target discovery to fe80::202:ff:fe00:319%eth0:3260 failed with status 0
[  230.740006] random: NetworkManager: uninitialized urandom read (16 bytes read)
[  230.846662] random: ln: uninitialized urandom read (6 bytes read)
iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
iscsiadm: connection login retries (reopen_max) 5 exceeded

We can find the iscsi discovery failure, and I believe it is caused by dracut iscsiroot.sh added the port number ":3260" string for IPv6 discovery.  BTW, I have checked the network is well.

**Comment by gonzoleeman** [comment]
If you believe dracut/iscsiroot.sh is causing the issues, try editing it and rebuilding initrd and trying it out.

Does this work for you with IPv4? I want to make sure we're chasing a dracut IPv6 issue.

**Comment by YanfeiXu** [comment]
> If you believe dracut/iscsiroot.sh is causing the issues, try editing it and rebuilding initrd and trying it out.
> 
> Does this work for you with IPv4? I want to make sure we're chasing a dracut IPv6 issue.

I was always using IPv4 for iscsiboot, it works.

Yes, I edited dracut/iscsiroot.sh to drop the ":3260" and retried. It still failed. Strange. 

dracut: ++ InitiatorName=xxxx-client.000200000318
dracut: + iscsi_initiator=xxxx-client.000200000318
dracut: + '[' -z xxxx-client.000200000318 ']'
dracut: + '[' -z '' ']'
dracut: + iscsi_target_port=3260
dracut: + '[' -z '' ']'
dracut: + iscsi_target_group=1
dracut: + '[' -z '' ']'
dracut: + iscsi_lun=0
dracut: + echo InitiatorName=xxxx-client.000200000318
dracut: + ln -fs /run/initiatorname.iscsi /dev/.initiatorname.iscsi
dracut: + '[' -e /etc/iscsi/initiatorname.iscsi ']'
dracut: + '[' -z '' ']'
dracut: + iscsid
dracut: + sleep 2
dracut: + '[' block:/dev/root = dhcp ']'
dracut: + '[' 'iscsi:[fe80::202:ff:fe00:319%eth0]::::iqn.e2000:acc' = dhcp ']'
dracut: ++ iscsiadm -m discovery -t st -p fe80::202:ff:fe00:319%eth0
dracut: ++ sed 's/^.*iqn/iqn/'
dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
dracut: iscsiadm: connection login retries (reopen_max) 5 exceeded
dracut: iscsiadm: Could not perform SendTargets discovery: iSCSI PDU timed out
dracut: + targets=
dracut: + '[' -z '' ']'
dracut: + echo 'Target discovery to fe80::202:ff:fe00:319%eth0 failed with status 0'
dracut: Target discovery to fe80::202:ff:fe00:319%eth0 failed with status 0

**Comment by YanfeiXu** [comment]
I thought I've got it. There was a fix for IPv6 discovery of dracut which comes after v051 (I am using v049). https://github.com/dracutdevs/dracut/commit/3b869230a08088a4f51d5bde8d5cfc8497779815
But it still using port number for IPv6. Anyway will try and put it back.

**Comment by YanfeiXu** [comment]
Still don't work.
-------log message--------
dracut: + '[' -n fe80::202:ff:fe00:319%eth0 -a -z fe80::202:ff:fe00:319%eth0 ']'
dracut: + iscsi_target_ip='[fe80::202:ff:fe00:319%eth0]'
dracut: ++ iscsiadm -m discovery -t st -p '[fe80::202:ff:fe00:319%eth0]'
dracut: ++ sed 's/^.*iqn/iqn/'
dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
[  271.419569] random: crng init done
dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out

**Comment by gonzoleeman** [comment]
On Mar 17, 2023, at 11:51 AM, Yanfei Xu ***@***.***> wrote:
> 
> 
> Still don't work.
> -------log message--------
> dracut: + '[' -n fe80::202:ff:fe00:319%eth0 -a -z fe80::202:ff:fe00:319%eth0 ']'
> dracut: + iscsi_target_ip='[fe80::202:ff:fe00:319%eth0]'
> dracut: ++ iscsiadm -m discovery -t st -p '[fe80::202:ff:fe00:319%eth0]'
> dracut: ++ sed 's/^.*iqn/iqn/'
> dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
> [ 271.419569] random: crng init done
> dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
> dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
> dracut: iscsiadm: connect to fe80::202:ff:fe00:319%eth0 timed out
> 

Give dracut the option to drop into the emergency shell, and try the command manually. Didn’t you say the exact same command worked for you, before, from a “normal” system? (You should verify that if not.)

What is the initiatorname you are using in your initrd? How is the ACL set up for your target? I’m no networking expert, but those messages almost make it look like the network isn’t up yet?

Again, the emergency shell will let you try it out for yourself and perhaps see why it isn’t working.

**Comment by YanfeiXu** [comment]
Hi Duncan, there indeed is a network issue. After solving it, I observed the "iscsiadm -m discovery" with interface name and without adding ":3260" worked:
###dracut: ++ iscsiadm -m discovery -t st -p fe80::203:ff:fe00:319%eth0
###dracut: ++ sed 's/^.*iqn/iqn/'
###dracut: + targets=iqn.e2000:acc
###dracut: + '[' -z iqn.e2000:acc ']'

But then failed at the next stage. I found it is caused by adding interface name. After I hacked dracut code to drop interface name at the second stage, It passed:
###dracut: + iscsiadm -m node -T iqn.e2000:acc -p fe80::203:ff:fe00:319 --op=update --name=node.startup --value=onboot --name=node.session.initial_login_retry_max --value=3

Unfortunately, with the early two stages passed, the third stage, login stage, failed:
###dracut: + iscsiadm -m node -L onboot
###dracut: iscsiadm: Could not login to [iface: default, target: iqn.e2000:acc, portal: fe80::203:ff:fe00:319,3260].
###dracut: iscsiadm: initiator reported error (8 - connection timed out)
###dracut: iscsiadm: Could not log into all portals

**Comment by gonzoleeman** [comment]
I edited the issue title, to reflect the actual problem better.

**Comment by YanfeiXu** [comment]
gentle ping, does this issue get confirmed or fixed?

**Comment by gonzoleeman** [comment]
You need to stop trying to boot from a link-local IPv6 address. They are not routable.

**Comment by MatthiasValvekens** [comment]
Note: this is completely untested, but the fix I had in mind was something like this: https://github.com/open-iscsi/open-iscsi/compare/master...MatthiasValvekens:open-iscsi:feature/ipv6-ll-discovery.

I'll see if I can find some time to figure out how to build an initramfs with this change included so I can validate whether this indeed works.
